### PR TITLE
Add apidsl file for toxencryptsave.

### PR DESCRIPTION
**Pull request by iphydf**
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/310)

### CODE REVIEW

**Comment by GrayHatter** [comment]
:lgtm:

**Comment by GrayHatter** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/310)*

File: toxencryptsave/toxencryptsave.h [hunk 1]
```diff
@@ -66,95 +67,113 @@ struct Tox_Options;
...
 typedef enum TOX_ERR_DECRYPTION {
+
+    /**
+     * The function returned successfully.
+     */
     TOX_ERR_DECRYPTION_OK,
+
     /**
-     * Some input data, or maybe the output pointer, was null.
+     * One of the arguments to the function was NULL when it was not expected.
```
Comment (zetok) [inline toxencryptsave/toxencryptsave.h]: When is NULL expected? Is it expected at all? Isn't this comment misleading?

File: toxencryptsave/toxencryptsave.h [hunk 2]
```diff
@@ -190,59 +222,66 @@ typedef struct {
  *
  * returns true on success
  */
-bool tox_derive_key_from_pass(const uint8_t *passphrase, size_t pplength, TOX_PASS_KEY *out_key,
-                              TOX_ERR_KEY_DERIVATION *error);
+bool tox_pass_key_derive_from_pass(struct Tox_Pass_Key *_key, const uint8_t *passphrase, size_t pplength,
```
Comment (zetok) [inline toxencryptsave/toxencryptsave.h]: >`tox_pass_key_derive_from_pass()`

Redundant department of redundant redundancy is calling. This function doesn't make much sense as far as the name goes.

File: toxencryptsave/toxencryptsave.h [hunk 3]
```diff
@@ -190,59 +222,66 @@ typedef struct {
...
+                                   TOX_ERR_KEY_DERIVATION *error);
 
-/* Same as above, except use the given salt for deterministic key derivation.
+/**
+ * Same as above, except use the given salt for deterministic key derivation.
  * The salt must be TOX_PASS_SALT_LENGTH bytes in length.
  */
-bool tox_derive_key_with_salt(const uint8_t *passphrase, size_t pplength, const uint8_t *salt, TOX_PASS_KEY *out_key,
-                              TOX_ERR_KEY_DERIVATION *error);
+bool tox_pass_key_derive_with_salt(struct Tox_Pass_Key *_key, const uint8_t *passphrase, size_t pplength,
```
Comment (zetok) [inline toxencryptsave/toxencryptsave.h]: >`*passphrase`

This is plain wrong.

File: toxencryptsave/toxencryptsave.h [hunk 4]
```diff
@@ -165,21 +184,34 @@ bool tox_pass_encrypt(const uint8_t *data, size_t data_len, const uint8_t *passp
...
+#ifndef TOX_PASS_KEY_DEFINED
+#define TOX_PASS_KEY_DEFINED
+typedef struct Tox_Pass_Key Tox_Pass_Key;
+#endif /* TOX_PASS_KEY_DEFINED */
+
+struct Tox_Pass_Key *tox_pass_key_new(void);
+
+void tox_pass_key_free(struct Tox_Pass_Key *_key);
+
+void tox_pass_key_get_key(const struct Tox_Pass_Key *_key, uint8_t *key);
```
Comment (zetok) [inline toxencryptsave/toxencryptsave.h]: No errors are possible? Should this actually read `TOX_PASS_KEY_LENGTH` bytes from given place in memory?
Why there are no docs on how to use it?
If it's something that shouldn't be used, why there's an API for it?

Same goes for other functions.

**Comment by iphydf** [comment]
Added to v0.1.0. I'll push it to v0.2.0 if we can't get it reviewed on time.

**Comment by zetok** [comment]
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[toxencryptsave/toxencryptsave.h, line 205 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/310#-KYqAtRswMPZ8-cPyYYp:-KYqAtRswMPZ8-cPyYYq:b-ybgstp) ([raw file](https://github.com/toktok/c-toxcore/blob/5ec295b6f19a95d0be4657bc7ffc2c6f9e2468e8/toxencryptsave/toxencryptsave.h#L205)):*
> ```objc
>  */
> bool tox_pass_encrypt(const uint8_t *plain, size_t plain_len, const uint8_t *passphrase, size_t passphrase_len,
>                       uint8_t *cipher, TOX_ERR_ENCRYPTION *error);
> ```

`cipher` as a name for output data is confusing, perhaps it should be renamed to e.g. `ciphertext`?
While at renaming, `plain` isn't that good either. Not sure if it could be renamed, given that `plaintext` isn't much more friendly (perhaps a bit).

Same goes for naming in other functions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/310)*

**Comment by iphydf** [comment]
Review status: 0 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[toxencryptsave/toxencryptsave.h, line 205 at r2](https://reviewable.io:443/reviews/toktok/c-toxcore/310#-KYqAtRswMPZ8-cPyYYp:-KYrsg4nTZ_5Y0l7fWWf:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/5ec295b6f19a95d0be4657bc7ffc2c6f9e2468e8/toxencryptsave/toxencryptsave.h#L205)):*
<details><summary><i>Previously, zetok wrote…</i></summary><blockquote>

`cipher` as a name for output data is confusing, perhaps it should be renamed to e.g. `ciphertext`?
While at renaming, `plain` isn't that good either. Not sure if it could be renamed, given that `plaintext` isn't much more friendly (perhaps a bit).

Same goes for naming in other functions.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/310)*

**Comment by robinlinden** [comment]
:lgtm_strong:

---

Reviewed 2 of 6 files at r2, 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[auto_tests/encryptsave_test.c, line 107 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/310#-KYsU1Sn8jp9re0Ld_mi:-KYsU1Sn8jp9re0Ld_mj:bl4uvzl) ([raw file](https://github.com/toktok/c-toxcore/blob/6c800b655e25609c9441c63cfc6268617ba57be5/auto_tests/encryptsave_test.c#L107)):*
> ```C
>     Tox_Pass_Key *key = tox_pass_key_new();
>     ck_assert_msg(key != NULL, "pass key allocation failure");
>     memcpy((uint8_t *)key, test_salt, 32);
> ```

`key->salt` and `key->key` may be clearer. If you disagree, then maybe at least use `TOX_PASS_SALT_LENGTH` instead of 32 when comparing and using offsets.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/310#-:-KYsV5JpB3YoL5U34IQ8:b-jbsf8z)*

**Comment by robinlinden** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/310)*

**Comment by iphydf** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[auto_tests/encryptsave_test.c, line 107 at r3](https://reviewable.io:443/reviews/toktok/c-toxcore/310#-KYsU1Sn8jp9re0Ld_mi:-KYsc-6l05jHP7BwHzBI:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/6c800b655e25609c9441c63cfc6268617ba57be5/auto_tests/encryptsave_test.c#L107)):*
<details><summary><i>Previously, robinlinden (Robin Lindén) wrote…</i></summary><blockquote>

`key->salt` and `key->key` may be clearer. If you disagree, then maybe at least use `TOX_PASS_SALT_LENGTH` instead of 32 when comparing and using offsets.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/310)*

**Comment by zetok** [comment]
Public API looks ok. :lgtm_strong:

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/310#-:-KYseZ6JvXt6HeUP1YA9:b-nre4g0)*

**Comment by iphydf** [comment]
@nurupo we can make modifications or undo this change if you have reasons for that later. The 0.1.0 release is still gated on your LGTM. I will merge the toxencryptsave API soon.

**Comment by nurupo** [comment]
Ok. I will have some reviewing time in 6 hours.

**Comment by nurupo** [comment]
Reviewed 1 of 6 files at r2, 3 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[toxencryptsave/toxencryptsave.h, line 51 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/310#-KYwTLCHbV_EVbgdBquI:-KYwTLCHbV_EVbgdBquJ:bn89b1o) ([raw file](https://github.com/toktok/c-toxcore/blob/4cf69996cc16282b13615843febd68ff6355eb7d/toxencryptsave/toxencryptsave.h#L51)):*
>\<unresolved\>

---

*[toxencryptsave/toxencryptsave.h, line 263 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/310#-KYwWUrpUA16Scb_B1bS:-KYwWUrpUA16Scb_B1bT:b8rqhqa) ([raw file](https://github.com/toktok/c-toxcore/blob/4cf69996cc16282b13615843febd68ff6355eb7d/toxencryptsave/toxencryptsave.h#L263)):*
> ```objc
>  */
> void tox_pass_key_free(struct Tox_Pass_Key *_key);
> ```

Why `_key`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/310)*

**Comment by GrayHatter** [comment]
Reviewed 2 of 6 files at r2, 3 of 4 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[toxencryptsave/toxencryptsave.api.h, line 54 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/310#-KYwlM6zYdC0ud5PjZos:-KYwlM6zYdC0ud5PjZot:blx48lc) ([raw file](https://github.com/toktok/c-toxcore/blob/4cf69996cc16282b13615843febd68ff6355eb7d/toxencryptsave/toxencryptsave.api.h#L54)):*
> ```objc
>  * ${pass_Key.derive}, then use the derived key to encrypt the data.
>  *
>  * The encrypted data is prepended with a magic number, to aid validity
> ```

Is the term here not cookie?

A magic number is something that makes things work, where why could easily unknown.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/310)*

**Comment by nurupo** [comment]
*[toxencryptsave/toxencryptsave.api.h, line 54 at r4](https://reviewable.io:443/reviews/toktok/c-toxcore/310#-KYwlM6zYdC0ud5PjZos:-KZ-J0gyCTSpvXJLshk9:b-4de69x) ([raw file](https://github.com/toktok/c-toxcore/blob/4cf69996cc16282b13615843febd68ff6355eb7d/toxencryptsave/toxencryptsave.api.h#L54)):*
<details><summary><i>Previously, GrayHatter (Gregory Mullen) wrote…</i></summary><blockquote>

Is the term here not cookie?

A magic number is something that makes things work, where why could easily unknown.
</blockquote></details>

Magic number is the correct term here https://en.wikipedia.org/wiki/Magic_number_(programming).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/toktok/c-toxcore/310)*

**Pull request merged**